### PR TITLE
Upgrade to ES6 classes and silence all warnings

### PR DESCRIPTION
**Pull request by MikeRatcliffe**
Associated Issue: #772

### Summary of Changes

* Always prefer ES6 classes in eslint.
* Switch all components over to ES6 classes.
* Use prop-types module instead of React.PropTypes.
* Use react-dom-factories instead of React.DOM.
* Upgraded dependencies that were causing warnings.

### CODE REVIEW

File: packages/devtools-launchpad/assets/Svg.js [hunk 1]
```diff
@@ -1,5 +1,5 @@
 const React = require("react");
-const InlineSVG = require("svg-inline-react");
+import InlineSVG from "svg-inline-react";
```
Comment (nchevobbe) [inline packages/devtools-launchpad/assets/Svg.js]: why should we do this ?
In my experience, `import` and `require` don't mix well.

File: packages/devtools-reps/package.json [hunk 2]
```diff
@@ -28,15 +28,17 @@
...
+    "svg-inline-react": "^3.0.0"
   },
   "devDependencies": {
     "babel-jest": "^20.0.3",
     "babel-plugin-transform-es2015-modules-commonjs": "^6.26.0",
     "babel-preset-react": "^6.24.1",
     "devtools-config": "^0.0.15",
-    "devtools-launchpad": "^0.0.101",
-    "devtools-license-check": "^0.2.0",
+    "devtools-launchpad": "^0.0.108",
```
Comment (nchevobbe) [inline packages/devtools-reps/package.json]: this may cause some issues, I need to check if the reps repl still work with this patch

File: packages/devtools-styleguide/src/components/Variables.js [hunk 3]
```diff
@@ -1,22 +1,18 @@
...
 function iterate(list, cbk) {
-  var index = list.length;
-  for (var i = 0; i < index; i++) {
+  let index = list.length;
+  for (let i = 0; i < index; i++) {
     cbk(list[i]);
   }
 }
 
-function map(list, cbk) {
```
Comment (nchevobbe) [inline packages/devtools-styleguide/src/components/Variables.js]: why was this removed ?

File: packages/devtools-launchpad/assets/Svg.js (hunk 1 above)
Comment (MikeRatcliffe) [reply packages/devtools-launchpad/assets/Svg.js]: To quote the author:

> I removed dist and es and made lib (which is ES2015 source) default, and there are several reason for it. For now use of ES2015 (and modules) is widespread, there are many tools supporting it (i.e. Webpack 2, Rollup), and you will use them anyway – **if you are using React/etc. If you want to stay in CommonJS land, please specify deps as 1.x.x. i.e) svg-inline-react: 1.x.x**

So require no longer works with this package.

The problem is that the 1.x.x versions trigger warnings so we have to use ES6 import.

File: packages/devtools-styleguide/src/components/Variables.js (hunk 3 above)
Comment (MikeRatcliffe) [reply packages/devtools-styleguide/src/components/Variables.js]: Because it is not used anywhere... we can put it back if you prefer.

File: .eslintrc [hunk 4]
```diff
@@ -67,7 +67,7 @@
     "react/no-unknown-property": "error",
     "react/prefer-es6-class": [
       "warn",
```
[the remark this answers is not in the document]
Comment (MikeRatcliffe) [reply .eslintrc]: We always want to use ES6 classes in this repo and I believe this .eslintrc is only used for this repo so we should be fine.

File: packages/devtools-components/src/tree.js [hunk 5]
```diff
@@ -138,7 +140,10 @@ const TreeNode = createFactory(createClass({
       renderItem(item, depth, focused, arrow, expanded)
     );
   }
-}));
+}
+
+const ArrowExpanderFactory = createFactory(ArrowExpander);
```
[the remark this answers is not in the document]
Comment (MikeRatcliffe) [reply packages/devtools-components/src/tree.js]: I didn't know that createFactory() is set to be deprecated (there is a TODO in the source).

Using new ArrowExpander() fails because you need a ReactClass and not a standard Class. The solution was to just replace createFactory with createElement calls, which is what the React folk recommend.

I have done this with all the crateFactory calls so that we don't need to change them in the future.

File: packages/devtools-components/src/tree.js [hunk 6]
```diff
@@ -249,165 +254,188 @@ function oncePerAnimationFrame(fn) {
...
-    return {
+  constructor(props) {
+    super(props);
+
+    this.state = {
       seen: new Set(),
     };
-  },
+
+    this._onExpand = oncePerAnimationFrame(this._onExpand);
```
[the remark this answers is not in the document]
Comment (MikeRatcliffe) [reply packages/devtools-components/src/tree.js]: Of course, done.

**Comment by brutalcrozt** [comment]
Yay it good news for me, it's really huge work ganbatte @MikeRatcliffe .  It have some conflict, i also waiting this one land.

**Comment by nchevobbe** [comment]
@MikeRatcliffe could you revert https://github.com/devtools-html/devtools-core/pull/775/commits/e09aec9512e512015eedf2d1f2a2c2bf14b66967 please ? We have a pending RFC about this, and we may not end doing it :)

**Comment by MikeRatcliffe** [comment]
Reverted :(

**Comment by nchevobbe** [comment]
This breaks the reps repl with: 
```
 Error: Module build failed: Error: No PostCSS Config found in: /Users/nchevobbe/Projects/devtools-core/packages/devtools-reps/src/reps
    at /Users/nchevobbe/Projects/devtools-core/packages/devtools-launchpad/node_modules/postcss-load-config/index.js:51:26
    at <anonymous>
```
This is due to devtools-launchpad bump in devtools-reps.
I will file a bug to fix this

**Comment by nchevobbe** [comment]
@MikeRatcliffe is this ready for review ?

**Comment by MikeRatcliffe** [comment]
@nchevobbe Yes, this is ready for review.

**Comment by nchevobbe** [comment]
sounds good, let's merge !
